site/content/entry/yuzu-progress-report-apr-2022/index.md

```diff
@@ -0,0 +1,332 @@
+Thanks to these changes, 'Super Smash Bros. Ultimate' no longer causes memory access issues while loading/unloading NROs, making the game stable for long runs.
+
+bunnei also {{< gh-hovercard "8013" "migrated slab heaps for the guest (Switch) kernel objects" >}} from host heap memory to emulated guest memory.
+With this change, yuzu's memory layout is now more closely matching the console.
+
+> A slab represents a contiguous piece of memory. A heap is a general term used for any memory that is
+allocated dynamically and randomly.
+
+So, slab heaps are basically pieces of memory dynamically allocated for guest kernel objects.
+By moving these away from the host (PC) heap memory (RAM) to emulated guest (Switch) memory, we can ensure
+that the kernel objects never go beyond the system limits to cause memory leaks on the host (PC).
+
+Thread local storage (TLS), the mechanism by which each thread in a given multithreaded process allocates storage for 
+thread-specific data, was also rewritten making it accurate to the latest HorizonOS behaviour.
+
+With these changes, we have now completely fixed the kernel memory object leaks that affected a few games, but went largely unnoticed, due to the previous implementation allowing unlimited allocations.
+
+Back to the list of April changes, bunnei also {{< gh-hovercard "8122" "reimplemented how yuzu handled thread allocation" >}} for HLE service interfaces.
+
+> Services are system processes running in the background which wait for incoming requests. 
+The Switch's HorizonOS has various services that perform various tasks e.g Audio, Bluetooth, etc.
+
+Previously we used to allocate one host thread per HLE service interface because -
+
+- some service routines need to block threads, and 
+- we don't support guest thread rescheduling from host threads.
+
+> A thread in block state will have to wait until an action can be completed.
+> exmaple needed
+
+The issue with this approach was that since it's the host OS that schedules these threads, yuzu could end up creating dozens of threads for services and that could lead to weird behaviour particularly on systems with hardware limitations.
+
+With the rewrite, yuzu now has a single "default service thread" that is used for 99% of the service methods that are non-blocking.
+For the services that are time-sensitive and for those that need blocking, we still allow thread creation (e.g. Audio, BSD, FileSystem, nvdrv)
+
+This brings down the service thread count from double digits to single digits, thus improving stability and consistency - especially on systems with less cores.
+Users with 4 thread CPUs (either 2 cores + HT/SMT, or 4 cores) should see performance and stability improvements on most games.
+
+Another battle for proper shutdown behaviour is fought and won.
+yuzu currently does not emulate multi-process capabilities of the HorizonOS kernel, however these still need to be managed.
```

Comment:
```suggestion
yuzu currently does not emulate multi-process capabilities of the HorizonOS kernel, because it is not required to emulate any games. However, the multi-process APIs that are used by games still need to be managed in the way they expect.
```